src/Grpc/HttpApi/src/Microsoft.AspNetCore.Grpc.HttpApi/Internal/Binding/HttpApiServiceMethodProvider.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Google.Protobuf.Reflection;
+using Grpc.AspNetCore.Server;
+using Grpc.AspNetCore.Server.Model;
+using Grpc.Shared;
+using Grpc.Shared.Server;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Grpc.HttpApi.Internal.Binding;
+
+internal sealed partial class HttpApiServiceMethodProvider<TService> : IServiceMethodProvider<TService> where TService : class
+{
+    private readonly ILogger<HttpApiServiceMethodProvider<TService>> _logger;
+    private readonly GrpcServiceOptions _globalOptions;
+    private readonly GrpcServiceOptions<TService> _serviceOptions;
+    private readonly GrpcHttpApiOptions _httpApiOptions;
+    private readonly ILoggerFactory _loggerFactory;
+    private readonly IServiceProvider _serviceProvider;
+    private readonly IGrpcServiceActivator<TService> _serviceActivator;
+
+    public HttpApiServiceMethodProvider(
+        ILoggerFactory loggerFactory,
+        IOptions<GrpcServiceOptions> globalOptions,
+        IOptions<GrpcServiceOptions<TService>> serviceOptions,
+        IServiceProvider serviceProvider,
+        IGrpcServiceActivator<TService> serviceActivator,
+        IOptions<GrpcHttpApiOptions> httpApiOptions)
+    {
+        _logger = loggerFactory.CreateLogger<HttpApiServiceMethodProvider<TService>>();
+        _globalOptions = globalOptions.Value;
+        _serviceOptions = serviceOptions.Value;
+        _httpApiOptions = httpApiOptions.Value;
+        _loggerFactory = loggerFactory;
+        _serviceProvider = serviceProvider;
+        _serviceActivator = serviceActivator;
+    }
+
+    public void OnServiceMethodDiscovery(ServiceMethodProviderContext<TService> context)
+    {
+        var bindMethodInfo = BindMethodFinder.GetBindMethod(typeof(TService));
+
+        // Invoke BindService(ServiceBinderBase, BaseType)
+        if (bindMethodInfo != null)
+        {
+            // The second parameter is always the service base type
+            var serviceParameter = bindMethodInfo.GetParameters()[1];
```

Comment:
Do we need to do a check that `GetParameters` returns an array with at least two items?

Reply:
No. Parameters are explicitly specified when getting the MethodInfo.